exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,384 @@
+				//  sophisticated error handler mechanism for this ForEach method.
+				forEachError = err
+				return
+			}
+
+			c, err := e.getCounter(record, mKey)
+			if err != nil {
+				// TODO: handle this better when we have a more
+				//  sophisticated error handler mechanism for this ForEach method.
+				forEachError = err
+				return
+			}
+
+			desc := record.Descriptor()
+			c.Add(value.CoerceToFloat64(desc.NumberKind()))
+
+			return
+		}
+
+		if gauge, ok := agg.(aggregator.LastValue); ok {
+			lv, _, err := gauge.LastValue()
+			if err != nil {
+				// TODO: handle this better when we have a more
+				//  sophisticated error handler mechanism for this ForEach method.
+				forEachError = err
+				return
+			}
+
+			g, err := e.getGauge(record, mKey)
+			if err != nil {
+				// TODO: handle this better when we have a more
+				//  sophisticated error handler mechanism for this ForEach method.
+				forEachError = err
+				return
+			}
+
+			desc := record.Descriptor()
+			g.Set(lv.CoerceToFloat64(desc.NumberKind()))
+
+			return
+		}
+	})
+
+	return forEachError
+}
+
+func (e *Exporter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	e.handler.ServeHTTP(w, r)
+}
+
+func (e *Exporter) Encode(kvs []core.KeyValue) string {
+	return e.labelEncoder.Encode(kvs)
+}
+
+func (e *Exporter) getCounter(record export.Record, mKey metricKey) (prometheus.Counter, error) {
+	e.Lock()
+	defer e.Unlock()
+	if c, ok := e.counters[mKey]; ok {
+		return c, nil
+	}
+
+	desc := record.Descriptor()
+	counterVec, err := e.getCounterVec(desc, record.Labels())
+	if err != nil {
+		return nil, err
+	}
+
+	counter, err := counterVec.GetMetricWithLabelValues(labelsToTags(record.Labels())...)
+	if err != nil {
+		return nil, err
+	}
+
+	e.counters[mKey] = counter
+	return counter, nil
+}
+
+func (e *Exporter) getGauge(record export.Record, mKey metricKey) (prometheus.Gauge, error) {
+	e.Lock()
+	defer e.Unlock()
+	if g, ok := e.gauges[mKey]; ok {
+		return g, nil
+	}
+
+	desc := record.Descriptor()
+	gaugeVec, err := e.getGaugeVec(desc, record.Labels())
+	if err != nil {
+		return nil, err
+	}
+
+	gauge, err := gaugeVec.GetMetricWithLabelValues(labelsToTags(record.Labels())...)
+	if err != nil {
+		return nil, err
+	}
+	e.gauges[mKey] = gauge
+	return gauge, nil
+}
+
+func (e *Exporter) getHistogram(record export.Record, mKey metricKey) (prometheus.Observer, error) {
+	e.Lock()
+	defer e.Unlock()
+	if g, ok := e.histograms[mKey]; ok {
+		return g, nil
+	}
+
+	desc := record.Descriptor()
+	histogramVec, err := e.getHistogramVec(desc, record.Labels())
+	if err != nil {
+		return nil, err
+	}
+
+	histogram, err := histogramVec.GetMetricWithLabelValues(labelsToTags(record.Labels())...)
+	if err != nil {
+		return nil, err
+	}
+	e.histograms[mKey] = histogram
+	return histogram, nil
+}
+
+func (e *Exporter) getCounterVec(desc *export.Descriptor, labels export.Labels) (*prometheus.CounterVec, error) {
+	if c, ok := e.counterVecs[desc]; ok {
```

Comment:
maybe a separate loc for the `*Vec` methods?

Reply:
Sorry, I actually didn't understand your suggestion here, could you discuss a little further?